subsys/bluetooth/controller/ll_sw/ull_adv_aux.c

```diff
@@ -777,34 +784,42 @@ uint8_t ull_adv_aux_hdr_set_clear(struct ll_adv_set *adv,
 	/* ADI */
 	{
 		struct pdu_adv_adi *pri_adi, *sec_adi;
-		uint16_t did = UINT16_MAX;
-
-		pri_dptr -= sizeof(struct pdu_adv_adi);
-		sec_dptr -= sizeof(struct pdu_adv_adi);
-
-		pri_adi = (void *)pri_dptr;
-		sec_adi = (void *)sec_dptr;
+		uint16_t did;
 
 		if (pri_hdr_prev.adi) {
 			struct pdu_adv_adi *pri_adi_prev;
 
 			pri_dptr_prev -= sizeof(struct pdu_adv_adi);
 			sec_dptr_prev -= sizeof(struct pdu_adv_adi);
 
-			memcpy(pri_dptr, pri_dptr_prev,
-			       sizeof(struct pdu_adv_adi));
-			memcpy(sec_dptr, sec_dptr_prev,
-			       sizeof(struct pdu_adv_adi));
-
 			pri_adi_prev = (void *)pri_dptr_prev;
 			did = sys_le16_to_cpu(pri_adi_prev->did);
-		} else {
-			pri_adi->sid = adv->sid;
-			sec_adi->sid = adv->sid;
+
+			/* Prevent using same DID if did_unique rolled over */
+			if (did == (did_unique & BIT_MASK(12))) {
+				did_unique++;
+			}
 		}
 
-		did++;
+		/* NOTE: Using a running did_unique is a simple solution to
+		 *       prevent using same DID when PDU fields change.
+		 *       This does not prevent a case wherein an advertising
+		 *       set is removed and added again while did_unique was
+		 *       incremented by other advertising sets, and there is a
+		 *       chance the new advertising set may end up with DID
+		 *       that was used before the set was removed. This is a
+		 *       rare case, which we are not handling.
+		 */
```

Comment:
I suppose the reason why we are not handling this, is because the solution to that would require a loop of all existing sets each time we increment the DID, which can be fairly expensive, and that the consequence of having two advertiser advertising with the same DID is not that bad? 

I don't know what the value is used for, so I don't know what the consequence of that is. 

In the rare case that DID is not unique, will it be non-unique for the lifetime of the advertising set, or just for the next advertising event? In order words, if it is not unique, will this be fixed over time automatically?

Reply:
> I suppose the reason why we are not handling this, is because the solution to that would require a loop of all existing sets each time we increment the DID, which can be fairly expensive, and that the consequence of having two advertiser advertising with the same DID is not that bad?

A correct solution option is, maintain an array of last used DIDs for each of maximum 16 SIDs that an application can use to create a set and use these values to set unique DID for PDU updates in a advertising set. Only expensive thing I can see is the setup time due to the requirement of 32 bytes of random value which depend on environmental conditions on how fast they are available.

I will post an update and we can review it.

Reply:
> I will post an update and we can review it.

Please do :) 

> If a scanner is using duplicate filter, then if will drop PDUs with same DID from being reported to Host.

That sounds like something that can, and should be, found during conformance tests, and as such is something we probably need to ensure doesn't happen.

Reply:
> That sounds like something that can, and should be, found during conformance tests, and as such is something we probably need to ensure doesn't happen.

Haven't observed that there was any conformance test checking duplicate filters, or the tests passed as old implementation did increment DID in normal cases.